interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,588 @@
+                    alpha_prog.append(ring_programs.momentum_compaction(
+                        0, order=order))
+
+            alpha_prog = ring_programs.momentum_compaction.combine_single_sections(
+                *alpha_prog, interpolation='linear')
+
+            setattr(self, alpha_name, alpha_prog.reshape(
+                self.n_sections, self.cycle_time, self.use_turns))
+
+        # Slippage factor derived from alpha, beta, gamma
+        for order in range(self.eta_orders + 1):
+            setattr(self, 'eta_%d' % (order), np.zeros(self.momentum.shape))
+        self._eta_generation()
+
+    @classmethod
+    def direct_input(cls, Particle, length, alpha_0,
+                     momentum=None, kin_energy=None, energy=None,
+                     bending_field=None, bending_radius=None, orbit_bump=None,
+                     alpha_1=None, alpha_2=None, **kwargs):
+
+        # Getting the number of sections for the length list
+        length = np.array(length, ndmin=1, dtype=float)
+        n_sections = len(length)
+
+        # Checking that at least one synchronous data input is passed
+        syncDataTypes = ('momentum', 'kin_energy', 'energy', 'B_field')
+        syncDataInput = (momentum, kin_energy, energy, bending_field)
+        assrt.single_not_none(*syncDataInput,
+                              msg='Exactly one of ' + str(syncDataTypes) +
+                              ' must be declared',
+                              exception=excpt.InputError)
+
+        # Taking the first synchronous_data input not declared as None
+        # The assertion above ensures that only one is declared
+        for func_type, synchronous_data in zip(syncDataTypes, syncDataInput):
+            if synchronous_data is not None:
+                break
+
+        # Casting synchronous data to datatype
+        if not isinstance(synchronous_data,
+                          ring_programs._synchronous_data_program):
+            synchronous_data \
+                = ring_programs._synchronous_data_program._conversions[
+                    func_type](synchronous_data)
+
+        # Casting orbit bump to datatype
+        if orbit_bump is not None:
+            if not isinstance(orbit_bump, ring_programs.orbit_length_program):
+                orbit_bump = ring_programs.orbit_length_program(orbit_bump)
+                print(orbit_bump.shape, type(orbit_bump))
+        else:
+            orbit_bump = [None] * n_sections
+
+        # Building all sections
+        Section_list = []
+        for index_section in range(n_sections):
+
+            # Passing sync_data as kwarg with the right func_type
+            sync_data = {func_type: synchronous_data[index_section]}
+
+            section = RingSection(
+                length[index_section], alpha_0, **sync_data,
+                orbit_bump=orbit_bump[index_section],
+                alpha_1=alpha_1, alpha_2=alpha_2,
+                **kwargs)
+
+            Section_list.append(section)
+
```

Comment:
Most of the data checking is a repeat of what is done in the `RingSection` objects.  Should be easy to just get everything into a list of the correct length and then construct all the sections, and let them handle the checking of stuff.  Something like:
```python
sync_data = {k, v for k, v in zip(syncDataTypes, syncDataInput)}
for k in sync_data:
    if sync_data[k] is None:
        sync_data[k] = [None]*n_sections

alphas = {'alpha_0': alpha_0, 'alpha_1': alpha_1, 'alpha_2': alpha_2, **{k: v for k, v in kwargs if 'alpha_' in k}}
for k in alphas:
    if alphas[k] is None:
        alphas[k] = [None]*n_sections
```
Then a quick length length:
```python
if not all([len(v) == n_sections for v in {**sync_data, **alphas}.values()])
    raise SomeError("Your data is crap")
```
Then the section can be made as:
```python
for i in range(n_sections):
    sectionInput = {**{k: sync_data[k][i] for k in sync_data}, **{k: alphas[k][i] for k in alphas}, **kwargs}
    Section_list.append(RingSection(**sectionInput)
```

I think something like this would reduce the amount of duplicate input testing and be easier to maintain if additional options are added.

Reply:
Noted, the direct_input is still work in progress I'll keep these in mind because indeed we need to minimize redundant code